pkg/hs/natpmp-static/hsrc_lib/Network/NATPMP.hsc

```diff
@@ -0,0 +1,260 @@
+
+#include <getgateway.h>
+#include <natpmp.h>
+#include <binding.h>
+
+import Prelude
+import Foreign
+import Foreign.C
+import Network.Socket
+
+import Control.Monad.IO.Unlift (MonadIO(..), MonadUnliftIO, withRunInIO)
+
+-- Opaque type for the internals of nat pmp
+data NatPmpStruct
+type NatPmpHandle = Ptr NatPmpStruct
+
+-- The response type, in its internal form. This struct is a C tagged union
+-- with additional data, but we need to read and write from its C form.
+data NatPmpResponse
+  = NatPmpResponsePublicAddress HostAddress
+  | NatPmpResponseUDPPortMapping Word16 Word16 Word32
+  | NatPmpResponseTCPPortMapping Word16 Word16 Word32
+  deriving (Show)
+
+instance Storable NatPmpResponse where
+  sizeOf    _ = #{size natpmpresp_t}
+  alignment _ = alignment (undefined :: CString)
+
+  peek p = do
+    t <- uintToEnum <$> (#{peek natpmpresp_t, type} p)
+    case t of
+      RTPublicAddress  ->
+        NatPmpResponsePublicAddress <$>
+          (#{peek natpmpresp_t, pnu.publicaddress.addr} p)
+      RTUDPPortMapping ->
+        NatPmpResponseUDPPortMapping
+          <$> (#{peek natpmpresp_t, pnu.newportmapping.privateport} p)
+          <*> (#{peek natpmpresp_t, pnu.newportmapping.mappedpublicport} p)
+          <*> (#{peek natpmpresp_t, pnu.newportmapping.lifetime} p)
+      RTTCPPortMapping ->
+        NatPmpResponseTCPPortMapping
+          <$> (#{peek natpmpresp_t, pnu.newportmapping.privateport} p)
+          <*> (#{peek natpmpresp_t, pnu.newportmapping.mappedpublicport} p)
+          <*> (#{peek natpmpresp_t, pnu.newportmapping.lifetime} p)
+
+  poke _ _ = error "Responses are an output data structure; poke makes no sense"
+
+type NatPmpResponseHandle = Ptr NatPmpResponse
+
+foreign import ccall unsafe "getgateway.h getdefaultgateway" _get_default_gateway :: Ptr CUInt -> IO CInt
+
+foreign import ccall unsafe "natpmp.h initnatpmp" _init_nat_pmp :: NatPmpHandle -> CInt -> CInt -> IO CInt
+foreign import ccall unsafe "natpmp.h closenatpmp" _close_nat_pmp :: NatPmpHandle -> IO CInt
+foreign import ccall unsafe "natpmp.h sendpublicaddressrequest" sendPublicAddressRequest :: NatPmpHandle -> IO CInt
+foreign import ccall unsafe "natpmp.h sendnewportmappingrequest" sendNewPortMappingRequest :: NatPmpHandle -> CInt -> CUShort -> CUShort -> CUInt -> IO CInt
+
+foreign import ccall unsafe "binding.h readNatResponseSynchronously" readNatResponseSynchronously :: NatPmpHandle -> NatPmpResponseHandle -> IO CInt
+
+-- Give the type system some help
+_peekCUInt :: Ptr CUInt -> IO CUInt
+_peekCUInt = peek
+
+uintToEnum :: Enum e => CUInt -> e
+uintToEnum = toEnum . fromIntegral
+
+intToEnum :: Enum e => CInt -> e
+intToEnum = toEnum . fromIntegral
+
+
+-- Fetches the default gateway as an ipv4 address
+getDefaultGateway :: IO (Maybe HostAddress)
+getDefaultGateway =
+  alloca $ \(pReturnAddr :: Ptr CUInt) -> do
+      ret <- _get_default_gateway pReturnAddr
+      case ret of
+        0 -> (Just . fromIntegral) <$> _peekCUInt pReturnAddr
```

Comment:
You get a compile error. This trick was learned from the lmdb bindings. _peekCUInt = peek, but with an explicit type declaration because the type system isn't smart enough to do the deduction.